File: html5_js/BreakBlock.js
/*******************************************************************************
	ブロック崩し
	参考: http://nn-hokuson.hatenablog.com/entry/2017/08/17/200918

	.NetCore/WPFで制作し、各クラス化とか速度の分解と角度の追加とか。
	ゲームとか殆ど作ったことがない人が作ったガラクタ。
	もう当たり判定全部短形でいいんじゃないかなとか。
	なんか斜めからボールが入ると当たり判定がガタガタになる。

	のhtml5/JavaScript版。
*******************************************************************************/

"use strict";

//ボールクラス
const Ball=(()=>{
	const draw=Symbol();
	const shps=Symbol();
	const deg=Symbol();
	const speed0=Symbol();
	const isStart=Symbol;
	const isCleared=Symbol();
	const color=Symbol();

	class Ball{
		get left(){return this.pos.X-this.radius;}
		get top(){return this.pos.Y-this.radius;}
		get right(){return this.pos.X+this.radius;}
		get bottom(){return this.pos.Y+this.radius;}

		constructor(_shps,radius,pos,_deg,speed,_color){
			this[shps]=_shps;
			this.radius=radius;
			this.pos0=pos;
			this[deg]=_deg;
			this[speed0]=speed;
			this[color]=_color;
			this.ready();
		}

		ready(){
			this.pos=Object.assign({},this.pos0);
			this.rad=this[deg]/180*Math.PI;
			this.speed=this[speed0];

			this[draw]();
			this[isStart]=false;
			setTimeout(()=>{this[isStart]=true},1500);
		}

		update(){
			//動くのは用意してからちょっと待つ
			if(!this[isStart]) return this[draw]();
			//ボール
			this.pos.X+=Math.cos(this.rad)*this.speed;
			this.pos.Y+=Math.sin(this.rad)*this.speed;
			//壁
			if(this.left<0 || Field.width<this.right) this.rad=Math.PI-this.rad;
			if(this.top<0 /*|| Field.height<this.bottom*/) this.rad=2*Math.PI-this.rad;
			//ボールロスト
			if(Field.height<this.top){
				this.ready();
				return;
			}

			this.rad%=2*Math.PI;
			this[draw]();
		}
	}

	Ball.prototype[draw]=function(){
		Field2d.beginPath();
		Field2d.arc(this.pos.X,this.pos.Y,this.radius,0,2*Math.PI,false);
		Field2d.fillStyle=this[color];
		Field2d.fill();
	};
	return Ball;
})();

//ブロッククラス
const Block=(()=>{
	const draw=Symbol();
	const lastHit=Symbol();
	const color=Symbol();
	const lines=Symbol();


	class Block{
		get left(){return this.pos.X-this.size.X/2;}
		get top(){return this.pos.Y-this.size.Y/2;}
		get right(){return this.pos.X+this.size.X/2;}
		get bottom(){return this.pos.Y+this.size.Y/2;}

		constructor(size,pos,_color){
			this.size=size;
			this.pos=pos;
			this[color]=_color;
		}

		update(){
			this[draw]();
		}
	}
	const pvt=Block.prototype;
	pvt[draw]=function(){
		Field2d.fillStyle=this[color];
		Field2d.fillRect(this.left,this.top,this.size.X,this.size.Y);
	};

	//上下左右の線のセット
	pvt.__defineGetter__[lines]=[
		[{X:this.left ,Y:this.top},   {X:this.right,Y:this.top}],
		[{X:this.right,Y:this.top},   {X:this.right,Y:this.bottom}],
		[{X:this.right,Y:this.bottom},{X:this.left, Y:this.bottom}],
		[{X:this.left ,Y:this.bottom},{X:this.left, Y:this.top}]
	];

	//ないせき
	const dot=(a,b)=>a.X*b.X+a.Y*b.Y;

	//線と円の当たり判定
	//理解できていないメソッド
	function lineVsCircle(p,center,radius){
		/*const lineDir=p[1]-p[0];                   //パドルの方向ベクトル
		Vector n=new Vector(lineDir.Y, -lineDir.X); //パドルの法線
		n.Normalize();

		Vector dir1=center-p[0];
		Vector dir2=center-p[1];

		double dist=Math.Abs(dir1*n);
		double a1=dir1*lineDir;
		double a2=dir2*lineDir;

		return a1*a2<0 && dist<radius;*/
	}
	return Block;
})();

//パドルクラス キーボードで動かせる
const Paddle=Block;

//ボールフィールドをグローバルのに保持
var Field,Field2d;

class MainWindow{
	static MainWindow(){
		Field=document.getElementById("Field");
		Field2d=Field.getContext("2d");

		var shps=new Array();
		shps.push(new Paddle({X:800,Y:5},{X:Field.width/2,Y:Field.height-50},"#9999FF",20));
		for(var i=0;i<7;i++){
			for(var n=0;n<5;n++){
				shps.push(new Block({X:80,Y:30},{X:50+i*90,Y:25+n*40},["#0000FF","#00FF00","#FF0000","#00FFFF","#FFFF00","#FF00FF","#FFFFFF"][i]));
			}
		}
		var ball=new Ball(shps,10,{X:Field.width/2,Y:300},220,10,"#FF00FF");

		//インターバル
		setInterval(()=>{
			Field2d.clearRect(0,0,Field.width,Field.height);
			for(var v of shps){
				v.update();
			}
			ball.update();
		},33);
		/*///Field2d.fillRect(0,0,100,100);
		Field2d.beginPath () ;
		Field2d.arc(100,100,50,0,2*Math.PI,false);
		// 塗りつぶしの色
		Field2d.fillStyle = "#3399FF";
		Field2d.fill() ;
		var i=0;
		function loop(){
			Field2d.beginPath();
			Field2d.clearRect(0,0,Field.width,Field.height);
			Field2d.arc(i,i,10,0,Math.PI,false);
			Field2d.fillStyle="#000099";
			Field2d.fill;
			i+=3;
		}
		//setInterval(loop,33);*/
	}
}

File: tkinter_py3/BreakBlock.py
"""*****************************************************************************
	ブロック崩し
	参考: http://nn-hokuson.hatenablog.com/entry/2017/08/17/200918

	.NetCore/WPFで制作し、各クラス化とか速度の分解と角度の追加とか。
	ゲームとか殆ど作ったことがない人が作ったガラクタ。
	もう当たり判定全部短形でいいんじゃないかなとか。
	なんか斜めからボールが入ると当たり判定がガタガタになる。

	のTkinter/Python3版。ゲームスピードはWPFより速く、HSPより遅い。
*****************************************************************************"""

import time
import threading
import math
from tkinter import messagebox as msg
from BBUtil import *

#ボールクラス
class Ball:
	@property
	def left(self):return self.pos.X-self.radius
	@property
	def top(self):return self.pos.Y-self.radius
	@property
	def right(self):return self.pos.X+self.radius
	@property
	def bottom(self):return self.pos.Y+self.radius

	def __init__(self,shps,radius,pos,deg,speed,color):
		self.__shps=shps
		self.radius=radius
		self.__pos0=pos
		self.__deg=deg
		self.__speed0=speed
		self.__isCleared=False
		self.__color=color
		self.__ready()

	def __ready(self):
		self.pos=self.__pos0.clone()
		self.__mpos=self.pos
		self.rad=self.__deg/180*math.pi
		self.speed=self.__speed0
		self.__ball=Field.create_oval(
			self.left,self.top,
			self.right,self.bottom,
			outline=self.__color,fill=self.__color
		)
		self.shp=self.__ball
		self.__draw()
		self.__isStart=threading.Thread(target=time.sleep,args=(1.500,))
		self.__isStart.start()

	def update(self):
		#クリア判定
		if len(self.__shps)<=1:
			self.__clear()
			return

		#動くのは用意してからちょっと待つ
		if self.__isStart.isAlive(): return
		#ボール
		self.__mpos=self.pos.clone()
		self.pos+=Vector(
			math.cos(self.rad)*self.speed,
			math.sin(self.rad)*self.speed
		);
		#壁
		if self.left<0 or float(Field.cget("width"))<self.right: self.rad=math.pi-self.rad
		if self.top<0: self.rad=2.0*math.pi-self.rad
		#ボールロスト
		if float(Field.cget("height"))<self.top:
			Field.delete(self.__ball)
			self.__ready()
			return

		#パドルとブロックの当たり判定
		def chkShps(v):
			if v.vsCircle(self.pos,self.radius):
				self.rad,self.speed=v.hit(self.rad,self.speed)
				if type(v) is not Block: return True
				Field.delete(v.shp)
				return False
			return True
		self.__shps=[v for v in self.__shps if chkShps(v)]
		self.rad%=2*math.pi
		self.__draw()

	def __draw(self):
		Field.move(self.__ball,self.pos.X-self.__mpos.X,self.pos.Y-self.__mpos.Y)

	#クリア処理
	def __clear(self):
		if self.__isCleared: return
		Field.delete(self.__ball)
		tb=Field.create_text(
			float(Field.cget("width"))/2,
			float(Field.cget("height"))/2,
			text="Clear!",
			fill="#AAAAFF",
			font=("メイリオ",72)
		)
		self.__isCleared=True

#ブロッククラス
class Block:
	@property
	def left(self):return self.pos.X-self.size.X/2
	@property
	def top(self):return self.pos.Y-self.size.Y/2
	@property
	def right(self):return self.pos.X+self.size.X/2
	@property
	def bottom(self):return self.pos.Y+self.size.Y/2

	def __init__(self,size,pos,color):
		self.size=size
		self.pos=pos

		self._block=Field.create_rectangle(
			self.left,self.top,
			self.right,self.bottom,
			outline=color,fill=color
		)
		self.shp=self._block

	def update(self):
		self._mpos=self.pos.clone()
		self._draw()

	def _draw(self):
		Field.move(self._block,self.pos.X-self._mpos.X,self.pos.Y-self._mpos.Y)

	#上下左右の線のセット
	@property
	def __lines(self): return [
		[Vector(self.left, self.top),   Vector(self.right,self.top)],
		[Vector(self.right,self.top),   Vector(self.right,self.bottom)],
		[Vector(self.right,self.bottom),Vector(self.left, self.bottom)],
		[Vector(self.left, self.bottom),Vector(self.left, self.top)]
	]

	#当たり判定
	def vsCircle(self,bPos,radius):
		for v in self.__lines:
			if Block.__lineVsCircle(v,bPos,radius):
				if v[0].X==v[1].X:
					if v[0].X==self.left: bPos.X=self.left-radius
					elif v[0].X==self.right: bPos.X=self.right+radius
				elif v[0].Y==v[1].Y:
					if v[0].Y==self.top: bPos.Y=self.top-radius
					elif v[0].Y==self.bottom: bPos.Y=self.bottom+radius
				self.__lastHit=v
				return True
		return False

	#線と円の当たり判定
	#理解できていないメソッド
	def __lineVsCircle(p,center,radius):
		lineDir=p[1]-p[0]              #パドルの方向ベクトル
		n=Vector(lineDir.Y,-lineDir.X) #パドルの法線
		n.Normalize()

		dir1=center-p[0]
		dir2=center-p[1]

		dist=abs(dir1*n)
		a1=dir1*lineDir
		a2=dir2*lineDir

		return a1*a2<0 and dist<radius

	#当たると反射する
	def hit(self,radB,speedB):
		x=self.__lastHit[1].X-self.__lastHit[0].X
		y=self.__lastHit[1].Y-self.__lastHit[0].Y
		line=1+math.cos(math.atan2(y,x))
		radB=line*math.pi-radB
		return radB,speedB

#パドルクラス キーボードで動かせる
class Paddle(Block):
	def __init__(self,size,pos,color,accel):
		super().__init__(size,pos,color)
		self.rad=0
		self.accel=accel
		self.speed=0

	#うごく
	def update(self):
		self._mpos=self.pos.clone()
		if CursorKey.left() and 0<self.left:
			self.speed=self.accel
			self.rad=math.pi
			self.pos.X-=self.speed
		elif CursorKey.right() and self.right<float(Field.cget("width")):
			self.speed=self.accel
			self.rad=0
			self.pos.X+=self.speed
		else: self.speed=0
		self._draw()

	#パドルのスピードでボールに変化をもたらす存在
	def hit(self,radB,speedB):
		(radB,speedB)=super().hit(radB,speedB)
		return radB+math.cos(self.rad)*self.speed/180*math.pi,speedB

#メインクラス
class MainWindow:
	def MainWindow(root,field):
		def printKey(event):
			key = event.keysym
			msg.showinfo("",key)
		global Field
		Field=field
		shps=[]
		shps.append(Paddle(Vector(100,5),Vector(float(Field.cget("width"))/2,float(Field.cget("height"))-50),"#66CCFF",20))
		blockColors=["#9999FF","#99FF99","#FF9999","#99FFFF","#FFFF99","#FF99FF","#FFFFFF"]
		shps.extend([Block(Vector(80,30),Vector(50+i*90,25+n*40),blockColors[i]) 
			for i in range(7)
				for n in range(5)])
		ball=Ball(shps,10,Vector(float(Field.cget("width"))/2,300),90,10,"#FF00FF")
		#インターバル
		def update():
			for v in shps:
				v.update()
			ball.update()
			root.after(33,update)
		update()
